subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -183,7 +183,11 @@ private synchronized void startWatching() {
                 @Override
                 public void handleChange(FileWatcherRegistry.Type type, Path path) {
                     try {
-                        fileEvents.put(new FileEvent(path, type));
+                        if (fileEvents.remainingCapacity() > 1) {
+                            fileEvents.put(new FileEvent(path, type));
+                        } else {
+                            fileEvents.offer(new FileEvent(true));
+                        }
```

Comment:
I don't understand what we are trying to accomplish here.

- `new FileEvent(true)` – can we replace this (and perhaps the other constructor as well) with a factory method so we can give it a name? Like, `FileEvent.signalLostState()`? Or maybe `LostStateEvent.instance()`?
- it seems that if the queue is full, we'd lose the lost event state. Why don't we `clear()` the queue instead? That'd remove the check for the queue size, too, which seems like a code smell to me. Any handling of the remaining events in the queue would be lost anyway, no?

Reply:
I added factory methods for the two type of events.

Moreover, when the queue is full we clear it, log a warning and put a lost state event in the queue.